File: 백준/Gold/9663. N－Queen/N－Queen.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var (
	reader  = bufio.NewReader(os.Stdin)
	writer  = bufio.NewWriter(os.Stdout)
	cnt     int
	visited []float64
)

func check(x int) bool {
	for i := 0; i < x; i++ {
		if visited[x] == visited[i] || math.Abs(visited[x]-visited[i]) == float64(x-i) {
			return false
		}
	}
	return true
}
func putQueen(x, n int) {
	if x == n {
		cnt += 1
	} else {
		for i := 0; i < n; i++ {
			visited[x] = float64(i)
			if check(x) {
				putQueen(x+1, n)
			}
		}
	}
}
func main() {
	defer writer.Flush()
	var n int
	fmt.Fscanf(reader, "%d\n", &n)
	cnt = 0
	visited = make([]float64, n)
	putQueen(0, n)
	fmt.Fprintf(writer, "%d\n", cnt)
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
input = sys.stdin.readline
cnt = 0

def check(x): #퀸 놓을 수 있는지 확인
    for i in range(x):
        if visited[x] == visited[i] or abs(visited[x] - visited[i]) == x - i:
            return False
    return True


def queen(x, n):
    global cnt
    if x == n:
        cnt += 1
    else:
        for i in range(n):
            visited[x] = i
            if check(x):
                queen(x + 1, n)

n = int(input())
visited = [0 for _ in range(n)]
queen(0, n)
print(cnt)